cmd/subctl/upgrade.go

```diff
@@ -0,0 +1,308 @@
+		upgradeOperatorVersion = upgradeSubctlVersion
+	}
+
+	// Upgrade Broker if installed; role updates are part of Broker redeploy
+	brokerUpgraded, err := upgradeBroker(ctx, clusterInfo, status)
+	if err != nil {
+		return err
+	}
+
+	var repository string
+	var debug bool
+	var imageOverride map[string]string
+
+	if clusterInfo.Submariner != nil {
+		repository = clusterInfo.Submariner.Spec.Repository
+		imageOverride = clusterInfo.Submariner.Spec.ImageOverrides
+		debug = clusterInfo.Submariner.Spec.Debug
+	} else if clusterInfo.ServiceDiscovery != nil {
+		repository = clusterInfo.ServiceDiscovery.Spec.Repository
+		imageOverride = clusterInfo.ServiceDiscovery.Spec.ImageOverrides
+		debug = clusterInfo.ServiceDiscovery.Spec.Debug
+	} else {
+		// Nothing further to do
+		return nil
+	}
+
+	// If a Broker was upgraded in this context, the Operator has already been upgraded
+	if !brokerUpgraded {
+		// Upgrade Operator if deployed
+		if err := upgradeOperator(ctx, clusterInfo, repository, debug, imageOverride, status); err != nil {
+			return err
+		}
+	}
+
+	// We want to show the user a version; use the most specific one
+	logVersion := upgradeSubctlVersion
+	if upgradeOperatorVersion != "" {
+		logVersion = upgradeOperatorVersion
+	}
+
+	if upgradeSubmarinerVersion != "" {
+		logVersion = upgradeSubmarinerVersion
+	}
+
+	// Upgrade Submariner
+	if err := upgradeConnectivity(ctx, clusterInfo, logVersion, status); err != nil {
+		return err
+	}
+
+	// Upgrade Service discovery
+	return upgradeServiceDiscovery(ctx, clusterInfo, logVersion, status)
+}
+
+func upgradeBroker(ctx context.Context, clusterInfo *cluster.Info, status reporter.Interface) (bool, error) {
+	status.Start("Checking if the Broker is installed")
+	defer status.End()
+
+	brokerObj, brokerFound, err := getBroker(ctx, clusterInfo.RestConfig, constants.DefaultBrokerNamespace)
+	if err != nil {
+		return false, status.Error(err, "Error checking for the Broker")
+	}
+
+	if !brokerFound {
+		return false, nil
+	}
+
+	status.Start("Upgrading the Broker to %s", upgradeOperatorVersion)
+	options := &deploy.BrokerOptions{
+		ImageVersion:    upgradeOperatorVersion,
+		BrokerNamespace: brokerObj.Namespace,
+		BrokerSpec:      brokerObj.Spec,
+	}
+
+	err = deploy.Deploy(ctx, options, status, clusterInfo.ClientProducer)
+
+	return err == nil, status.Error(err, "Error upgrading the Broker")
+}
+
+func upgradeOperator(ctx context.Context, clusterInfo *cluster.Info, repository string, debug bool, imageOverride map[string]string,
+	status reporter.Interface,
+) error {
+	status.Start("Checking if the Operator is installed")
+	defer status.End()
+
+	operatorDeployment, err := clusterInfo.ClientProducer.ForKubernetes().AppsV1().Deployments(constants.OperatorNamespace).
+		Get(ctx, names.OperatorComponent, metav1.GetOptions{})
+	if err != nil && !errors.IsNotFound(err) {
+		return status.Error(err, "Error retrieving Operator deployment")
+	}
+
+	if operatorDeployment == nil {
+		return nil
+	}
```

Comment:
Is it guaranteed that `operatorDeployment` will be nil if not found? I think it's clearer to key off of `IsNotFound`.

```suggestion
	if errors.IsNotFound(err) {
		return nil
	}

	if err != nil  {
		return status.Error(err, "Error retrieving Operator deployment")
	}
```

Reply:
Right, given the implementation, probably not.